File: Python/algorithmic_heights/17_dij.py
# Dijkstra's Algorithm
import sys

from collections import defaultdict, namedtuple
from math import inf, isinf

Edge = namedtuple('Edge', ['node_from', 'node_to', 'weight'])


class DirectedGraph:
    def __init__(self, nr_nodes, edge_list):
        self._nr_nodes = nr_nodes
        self._adjacency_list = self._build_adjacency_list(edge_list)

    @staticmethod
    def _build_adjacency_list(edge_list):
        adjacency_list = defaultdict(list)
        for edge in edge_list:
            adjacency_list[edge.node_from].append(edge)
        return dict(adjacency_list)

    @property
    def nr_nodes(self):
        return self._nr_nodes

    def neighbours(self, node):
        return self._adjacency_list.get(node, [])


def convert_to_intlist(line):
    return tuple(int(item) for item in line.split())


def read_edges(reader, nr_edges):
    edges = []
    for _ in range(nr_edges):
        a, b, w = convert_to_intlist(next(reader))
        edges.append(Edge(node_from=a, node_to=b, weight=w))
    return edges


def update_distances(neighbours, dist_u, distances, nodes):
    for _, v, weight in neighbours:
        distance_through_u = dist_u + weight
        if distances[v-1] > distance_through_u:
            distances[v-1] = distance_through_u
            nodes[v] = distance_through_u


def calc_shortest_distances(graph, source_node):
    distances = [inf] * graph.nr_nodes
    distances[source_node-1] = 0
    nodes = {node: distances[node-1] for node in range(1, graph.nr_nodes+1)}
    while nodes:
         u = min(nodes, key=nodes.get)
         dist_u = nodes.pop(u)
         neighbours = graph.neighbours(u)
         update_distances(neighbours, dist_u, distances, nodes)

    return tuple(-1 if isinf(d) else d for d in distances)


def main():
    reader = sys.stdin
    nr_nodes, nr_edges = convert_to_intlist(next(reader))
    edge_list = read_edges(reader, nr_edges)
    graph = DirectedGraph(nr_nodes, edge_list)
    result = calc_shortest_distances(graph, source_node=1)
    print(' '.join(map(str, result)))


if __name__ == '__main__':
    main()

File: Haskell/algorithmic_heights/17_dij.hs
import Control.Monad (replicateM)
import Data.IntMap (IntMap)
import qualified Data.IntMap as M
import Data.Heap (MinPrioHeap)
import qualified Data.Heap as H
import Data.IntSet (IntSet)
import qualified Data.IntSet as S

data Edge = Edge { nodeFrom :: Int, nodeTo :: Int, weight :: Int }
type AdjacencyList = IntMap [Edge]
data Graph = Graph { nrNodes :: Int, adjacencyList :: AdjacencyList }

data Distance = Dist Int | Infinity deriving Eq
data State = State { queue :: MinPrioHeap Distance Int, distances :: IntMap Distance, finalizedNodes :: IntSet }


instance Ord Distance where
    Infinity <= Infinity = True
    Dist d1 <= Dist d2 = d1 <= d2
    Infinity <= Dist _ = False
    Dist _ <= Infinity = True


instance Show Distance where
    show (Dist d) = show d
    show Infinity = "-1"


addDist :: Distance -> Distance -> Distance
addDist (Dist d1) (Dist d2) = Dist (d1 + d2)
addDist _ _ = Infinity


convertToIntList :: String -> [Int]
convertToIntList = map read . words


readInputLine :: String -> (Int, Int)
readInputLine line =
    case convertToIntList line of
        [a, b] -> (a, b)
        _ -> error "Malformed input."


readEdges :: Int -> IO [Edge]
readEdges nrEdges = do
    let parse line =
            case convertToIntList line of
                [a, b, w] -> Edge a b w
                _ -> error "Malformed edge input."
    edges <- replicateM nrEdges getLine
    return $ map parse edges


createGraph :: Int -> [Edge] -> Graph
createGraph nrNodes edges =
    let addEdge adjacencyList edge =
            let update Nothing = Just [edge]
                update (Just ns) = Just (edge : ns)
            in M.alter update (nodeFrom edge) adjacencyList
        adjacencyList = foldl addEdge M.empty edges
    in Graph { nrNodes = nrNodes, adjacencyList = adjacencyList }


getNeighbors :: Graph -> Int -> [Edge]
getNeighbors (Graph _ adjacencyList) node = M.findWithDefault [] node adjacencyList


updateDistances :: State -> [Edge] -> Distance -> State
updateDistances state neighbors dist =
    let go currentState [] = currentState
        go currentState @ (State queue distances _) ((Edge _ v weight) : ns) =
            let distanceThroughNode = addDist dist (Dist weight)
                currentDistance = M.findWithDefault Infinity v distances
            in if currentDistance <= distanceThroughNode then go currentState ns
            else
                let queue' = H.insert (distanceThroughNode, v) queue
                    distances' =  M.insert v distanceThroughNode distances
                in go (currentState { queue = queue', distances = distances' }) ns
    in go state neighbors


calcShortestDistances :: Graph -> Int -> [Distance]
calcShortestDistances graph @ (Graph nrNodes adjacencyList) sourceNode =
    let go state @ (State queue distances finalizedNodes) =
            case H.view queue of
                Nothing -> map (\node -> M.findWithDefault Infinity node distances) [1..nrNodes]
                Just ((dist, node), rest) ->
                    if S.member node finalizedNodes then go (State rest distances finalizedNodes)
                    else
                        let neighbors = getNeighbors graph node
                            state' = updateDistances state neighbors dist
                        in go state' { finalizedNodes = S.insert node finalizedNodes }
    in go $ State (H.fromList [(Dist 0, sourceNode)]) ( M.singleton sourceNode (Dist 0)) S.empty


main :: IO ()
main = do
    (nrNodes, nrEdges) <- readInputLine <$> getLine
    edges <- readEdges nrEdges
    let graph = createGraph nrNodes edges
    let result = calcShortestDistances graph 1
    putStrLn $ unwords $ map show result
